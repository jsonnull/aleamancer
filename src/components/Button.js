// @flow
import styled, { css } from 'styled-components'
import * as themes from 'styles/themes'
import { fontSize, colors } from 'styles/common'

const redMixin = css`
  background-color: ${colors.red};
  color: white;
  &:hover {
    background-color: ${colors.red};
    color: white;
  }
`

const greenMixin = css`
  background-color: ${colors.green};
  color: white;
  &:hover {
    background-color: ${colors.green};
    color: white;
  }
`

const outlineMixin = css`
  background-color: transparent;
  border: 1px solid ${props => props.theme.color};
  color: ${props => props.theme.color};
  &:hover {
    background-color: transparent;
  }
`

export const Button = styled.button`
  display: inline-block;
  font-size: ${fontSize.small};
  line-height: 3rem;
  height: 3rem;
  padding: 0 1rem;
  outline: 0;
  color: ${themes.dark.color};
  background-color: ${colors.blue};
  border: none;
  border-radius: 4px;
  cursor: pointer;
  transition: background, text-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-out;
  flex: 1;

  &:hover {
    text-shadow: 0 0 3px white;
    background-color: ${colors.lightBlue};
    transition: background, text-shadow;
    transition-duration: 0.1s;
    transition-timing-function: ease-out;
  }

  ${props => (props.red ? redMixin : null)} ${props =>
      props.green ? greenMixin : null} ${props =>
      props.outline ? outlineMixin : null};
`

export default Button
