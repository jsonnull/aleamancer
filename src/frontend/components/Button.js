// @flow
import styled, { css } from 'styled-components'
import * as themes from 'frontend/styles/themes'
import { colors, fontSize, timings } from 'frontend/styles/common'

function shade(hex: string, lum: number = 0) {
  hex = hex.replace(/[^0-9a-f]/gi, '')

  console.assert(hex.length === 6)

  let rgb = '#'

  for (let i = 0; i < 3; i++) {
    let c = parseInt(hex.substr(i * 2, 2), 16)
    c = Math.round(Math.min(Math.max(0, c + c * lum), 255)).toString(16)
    rgb += ('00' + c).substr(c.length)
  }

  return rgb
}

const boxShadow = color => `inset 0 -20px 10px -10px ${shade(color, -0.06)}`
const hoverBackground = color => shade(color, -0.09)
const hoverBoxShadow = color =>
  `inset 0 -20px 10px -10px ${shade(color, -0.15)};`

const redMixin = css`
  background: ${colors.red};
  box-shadow: ${boxShadow(colors.red)};
  &:hover {
    background: ${hoverBackground(colors.red)};
    box-shadow: ${hoverBoxShadow(colors.red)};
  }
`

const greenMixin = css`
  background: ${colors.green};
  box-shadow: ${boxShadow(colors.green)};
  &:hover {
    background: ${hoverBackground(colors.green)};
    box-shadow: ${hoverBoxShadow(colors.green)};
  }
`

const outlineMixin = css`
  background: transparent;
  box-shadow: none;
  border: 1px solid ${props => props.theme.color};
  color: ${props => props.theme.color};
  &:hover {
    box-shadow: ${props => hoverBoxShadow(props.theme.background)};
    background: transparent;
  }
`

export const Button = styled.button`
  display: inline-block;
  font-size: ${fontSize.small};
  line-height: 3rem;
  height: 3rem;
  padding: 0 1rem;
  outline: 0;
  color: ${themes.dark.color};
  background: ${colors.blue};
  box-shadow: ${boxShadow(colors.blue)};
  border: none;
  border-radius: 4px;
  cursor: pointer;
  transition: background, box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ${timings.easeOutQuad};
  flex: 1;

  &:hover {
    background: ${hoverBackground(colors.blue)};
    box-shadow: ${hoverBoxShadow(colors.blue)};
    transition: background, box-shadow;
    transition-duration: 0.1s;
    transition-timing-function: ${timings.easeOutQuad};
  }

  ${props => (props.red ? redMixin : null)};
  ${props => (props.green ? greenMixin : null)};
  ${props => (props.outline ? outlineMixin : null)};
`

export default Button
