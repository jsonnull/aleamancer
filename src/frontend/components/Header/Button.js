// @flow
import styled from 'styled-components'
import { fontSize } from 'frontend/styles/common'

const Button = styled.div`
  height: 3.25rem;
  width: ${props => (props.square ? '3.25rem' : 'auto')};
  line-height: 3.25rem;
  padding: 0 1rem;
  text-align: center;
  font-size: ${fontSize.normal};
  background: ${props => props.theme.backgroundInverted};
  cursor: pointer;
  border-radius: 5px;
  color: ${props => props.theme.colorInverted};

  &:hover {
    background: ${props => props.theme.backgroundInvertedSecondary};
  }
`

export default Button
