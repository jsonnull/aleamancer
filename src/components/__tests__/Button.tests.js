// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import Button from '../Button.js'

describe('Button component', () => {
  it('renders correctly with no options', () => {
    const tree = renderer.create(<Button />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with red option set', () => {
    const tree = renderer.create(<Button red />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with green option set', () => {
    const tree = renderer.create(<Button green />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with outline option set', () => {
    const tree = renderer.create(<Button outline />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
