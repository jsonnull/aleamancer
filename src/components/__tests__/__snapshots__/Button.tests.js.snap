// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly with green option set 1`] = `
<button
  className="sc-bdVaJa jMuSlE"
/>
`;

exports[`Button component renders correctly with no options 1`] = `
<button
  className="sc-bdVaJa gXXZoz"
/>
`;

exports[`Button component renders correctly with outline option set 1`] = `
<button
  className="sc-bdVaJa giJqZv"
/>
`;

exports[`Button component renders correctly with red option set 1`] = `
<button
  className="sc-bdVaJa gyRotp"
/>
`;
